versioned_hdf5/wrappers.py

```diff
@@ -818,72 +818,82 @@ def compression_opts(self, value):
 class InMemoryArrayDataset(DatasetLike):
     """
     Class that looks like a h5py.Dataset but is backed by an array
+    with support for lazy allocation.
     """
 
     def __init__(
         self,
         name: str,
-        array: np.ndarray,
+        array: np.ndarray | None,
         *,
         parent: InMemoryGroup,
         fillvalue: Any | None = None,
         chunks: tuple[int, ...] | None = None,
+        shape: tuple[int, ...] | None = None,
+        dtype: DTypeLike | None = None,
     ):
         self.name = name
-        self._array = array
         self.attrs = {}
         self.parent = parent
         self._fillvalue = fillvalue
+
+        self._array = array
+        self._lazy_shape = array.shape if array is not None else shape
+        self._lazy_dtype = array.dtype if array is not None else np.dtype(dtype)
+        if self._array is None and (self._lazy_shape is None or self._lazy_dtype is None):
```

Comment:
array will never be None.
An InMemoryArrayDataset is created in two use cases:
1. a user calls `create_dataset` and passes an array as the `data=` parameter;
2. a user calls `stage_version` and then opens a pre-existing dataset, which creates a `InMemoryDataset`; then it calls `__setitem__` on it to update everything. `DatasetWrapper.__setitem__` hot-swaps the `InMemoryDataset` with a new `InMemoryArrayDataset`.